File: kamacoder47/kamacoder47.cpp
// 朴素dijkstra
#include<iostream>
#include<vector>
#include<climits>

using namespace std;

int main(){
    int n, m, p1, p2, val;
    cin >> n >> m;

    vector<vector<int>> grid(n + 1, vector<int>(n + 1, INT_MAX));
    for (int i = 0; i < m; i++){
        cin >> p1 >> p2 >> val;
        grid[p1][p2] = val;
    }

    int start = 1;
    int end = n;

    vector<int> minDist(n + 1, INT_MAX);
    vector<bool> visited(n + 1, false);

    minDist[start] = 0;
    for (int i = 1; i <= n; i++){
        int minVal = INT_MAX;
        int cur = 1;

        for (int v = 1; v <= n; v++){
            if (!visited[v] && minDist[v] < minVal){
                minVal = minDist[v];
                cur = v;
            }
        }

        visited[cur] = true;

        for (int v = 1; v <= n; v++){
            if (!visited[v] && grid[cur][v] != INT_MAX && minDist[cur] + grid[cur][v] < minDist[v]){
                minDist[v] = minDist[cur] + grid[cur][v];
            }
        }
    }

    if (minDist[end] == INT_MAX) cout << -1 << endl;
    else cout << minDist[end] << endl;
}

// dijkstra堆优化版本

#include<iostream>
#include<vector>
#include<list>
#include<queue>
#include<climits>

using namespace std;

class mycomparison{
public:
    bool operator()(const pair<int, int>& lhs, const pair<int, int>& rhs){
        return lhs.second > rhs.second;
    }
};

struct Edge
{
    int to;
    int val;
    Edge(int t, int w): to(t), val(w) {}
};

int main(){
    int n, m, p1, p2, val;
    cin >> n >> m;
    vector<list<Edge>> grid(n + 1);

    for (int i = 0; i < m; i++){
        cin >> p1 >> p2 >> val;
        grid[p1].push_back(Edge(p2, val));
    }

    int start = 1;
    int end = n;

    vector<int> minDist(n + 1, INT_MAX);

    vector<bool> visited(n + 1, false);

    priority_queue<pair<int, int>, vector<pair<int, int>>, mycomparison> pq;

    pq.push(pair<int, int>(start, 0));

    minDist[start] = 0;

    while (!pq.empty())
    {
        pair<int, int> cur = pq.top();
        pq.pop();

        if (visited[cur.first]) continue;

        visited[cur.first] = true;

        for (Edge edge : grid[cur.first]){
            if (!visited[edge.to] && minDist[cur.first] + edge.val < minDist[edge.to]){
                minDist[edge.to] = minDist[cur.first] + edge.val;
                pq.push(pair<int, int>(edge.to, minDist[edge.to]));
            }
        }
    }
    if (minDist[end] == INT_MAX) cout << -1 << endl;
    else cout << minDist[end] << endl;
}

File: kamacoder47/kamacoder47.py
# 朴素dijkstra算法
import sys
INT_MAX = sys.maxsize

n, m = map(int, input().split())
grip = [[INT_MAX] * (n + 1) for _ in range(n + 1)]

for i in range(m):
    p1, p2, val = map(int, input().split())
    grip[p1][p2] = val

start = 1
end = n

visited = [False] * (n + 1)
miniDist = [INT_MAX] * (n + 1)

miniDist[start] = 0

for i in range(1, n + 1):
    minVal = INT_MAX
    cur = 1
    for v in range(1, n + 1):
        if not visited[v] and miniDist[v] < minVal:
            minVal = miniDist[v]
            cur = v
    
    visited[cur] = True

    for v in range(1, n + 1):
        if not visited[v] and grip[cur][v] != INT_MAX and miniDist[cur] + grip[cur][v] < miniDist[v]:
            miniDist[v] = miniDist[cur] + grip[cur][v]
        

if miniDist[end] == INT_MAX:
    print(-1)
else:
    print(miniDist[end])


#  dijkstra堆优化版本

import heapq

class Edge:
    def __init__(self, to, val):
        self.to = to
        self.val = val

def dijkstra(n, m, edges, start, end):
    grid = [[] for _ in range(n + 1)]
    for p1, p2, val in edges:
        grid[p1][p2] = val
    
    miniDist = [float('inf')] * (n + 1)
    visited = [False] * (n + 1)

    pq = []
    heapq.heappush(pq, (0, start))
    miniDist[start] = 0

    while pq:
        cur_dist, cur_node = heapq.heappop(pq)

        if visited[cur_node]:
            continue
        visited[cur_node] = True

        for edge in grid[cur_node]:
            if not visited[edge.to] and cur_dist + edge.val < miniDist[edge.to]:
                miniDist[edge.to] = cur_dist + edge.val
                heapq.heappush(pq, (miniDist[edge.to], edge.to))
    return -1 if miniDist[end] == float('inf') else miniDist[end]

n, m = map(int, input().split())

edges = [tuple(map(int, input().split())) for _ in range(m)]
start = 1
end = n
result = dijkstra(n, m, edges, start, end)
print(result)